### PR TITLE
[perf] MiniMax-H3 LoRA training on a B200 node: 3.74 s to 2.47 s per step at 8 GPUs, mostly from compiling the DiT block body

### PR DESCRIPTION
**Pull request by TarzanZhao**
## Summary

MiniMax-H3 FL2VA LoRA training (`examples/minimax_h3/model_training/lora/MiniMax-H3-FL2VA.sh`, stage 2) spends a large part of each step outside the GEMMs and attention. The element-wise body of the 50 DiT blocks runs as about 600 kernel launches per block execution, `cu_seqlens.tolist()` forces a device sync in every block execution (102 per step once gradient checkpointing recomputes each block), and the checkpoint recomputes the whole block including its attention. This PR compiles the DiT block body for training and moves the checkpoint inside the compiled function with a policy that keeps the attention output. On the 8 × B200 node I measured, the median step goes from 3.74 s to 2.47 s at 8-process DDP, **1.51x**; I have not measured any other GPU.

Both commits are in `diffsynth/models/minimax_h3_dit.py` and go through neither `compile_pipeline()` nor `diffsynth/core/gradient/gradient_checkpoint.py`. The compiled functions and the new switch are module-level, so they cover every MiniMax-H3 DiT in the process, not one pipeline instance the way `compile_pipeline()` does.

1. **Compile the DiT block body for training, one `cu_seqlens` sync per forward** (`21bafaf`). `MiniMaxH3DiTBlock.forward` now calls one module-level `torch.compile(dynamic=False)` function shared by the 50 blocks when grad is enabled and CUDA is available, fusing RoPE, the six AdaLN `index_select` gathers, RMSNorm, SwiGLU and the residual gates; `MiniMaxH3DiT.forward` converts `cu_seqlens` once and passes the bounds down as Python ints. Inference keeps the eager path. `dynamic=False` with the bounds as ints specialises the graph on the exact segment bounds, so any other packed shape compiles again and past Dynamo's recompile limit the block falls back to eager with only a Dynamo warning — I measured one cached sample, one shape for all 30 steps. `DIFFSYNTH_COMPILE_DIT=0` turns the compile off, but it is on by default, which makes it the only `DIFFSYNTH_*` variable whose unset value is the *new* behaviour; this PR also does not add it to `docs/{en,zh}/Pipeline_Usage/Environment_Variables.md`. Say which you would prefer and I will flip the default, document it, or both.
2. **Selective activation checkpointing, attention output kept** (`262afc0`). With the body compiled, the checkpoint recompute is the next largest cost, and the attention forward is its most expensive op relative to the memory its output holds, about 240 MB per block in bf16. The checkpoint moves inside the compiled function under a policy that saves `aten::_scaled_dot_product_{cudnn,flash,efficient}_attention` and recomputes the rest; that is what raises peak memory below. It has no switch of its own and runs only when `use_gradient_checkpointing` is on without offload and the compile is on: the offload and DeepSpeed paths keep `gradient_checkpoint_forward`, and `DIFFSYNTH_COMPILE_DIT=0` drops this too. Under an attention dispatch that does not call those three ops, flash-attn for instance, the policy matches nothing and the block recomputes everything again.

Not exercised at all: the ZeRO-3 full-parameter recipe, FSDP, the ascend backend, inference, resolutions or frame counts other than the one below, GPUs other than B200, torch other than 2.14.0.

### Speedup Result

Measured on one node of 8 × B200, two interleaved runs per version; each cell holds both runs.

| | baseline (`ce9f454`, `DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch`) | this PR | change |
|---|---|---|---|
| step time, median of steps 6 to 30, rank 0 | 3.7422 / 3.7433 s | 2.4737 / 2.4728 s | **1.51x**, −33.9 % |
| step time, mean of steps 6 to 30 | 3.7493 / 3.7522 s | 2.4732 / 2.4725 s | |
| peak memory per rank (nvidia-smi `memory.used`) | 81,408 MiB | 91,610 MiB | +10.2 GB, +12.5 % |
| step 1 | 7.36 / 7.25 s | 11.81 / 11.34 s with a warm inductor cache, 34.8 s cold | one-time compile |

| commit | measured effect |
|---|---|
| `21bafaf`, compiled block body and one `cu_seqlens` sync per forward | 3.802 s to 2.866 s |
| `262afc0`, selective activation checkpointing | 2.866 s to 2.530 s, −11.7 % |

Those two rows come from the earlier version of this chain, not from the baseline above: another 8 × B200 node, `DIFFSYNTH_ATTENTION_IMPLEMENTATION` unset, on top of a cuDNN-SDPA commit no longer in this branch. I did not measure the commits one at a time on the baseline above.

## Correctness Verification

I ran the unmodified `ce9f454` and this branch on the same data with the same setup and seeds, 30 steps on each of 8 ranks, and compared 5,040 recorded values against tolerances set before the runs, at about three times the spread of three runs of the unmodified code. Nothing exceeds its tolerance.

One of the recorded quantities is weak evidence and I would rather say so than let the pass rate speak for it: the gradient norm over all 416 LoRA tensors differs by as much as **4.8x the baseline value** (0.043 against 0.247, step 14, one rank). It passes only because that check's budget is 20x — it was written to catch NaNs and blow-ups, not to bound drift. The rows that do bound drift are the other seven, and the closest of those is at 0.51 of its budget.

| recorded | baseline vs this PR | tolerance |
|---|---|---|
| mean absolute value per LoRA A tensor after step 30, 1,664 tensors over 8 ranks | 0.95 % max | 2 % |
| mean absolute value per LoRA B tensor after step 30, 1,664 tensors, values around 5e-4 | 1.4e-4 max | 4e-4 absolute |
| loss, 30 steps × 8 ranks | 2.4 % max, 0.0076 absolute | 15 % or 0.03 absolute |
| DiT video output, mean of \|noise_pred\|, 30 steps × 8 ranks | 0.41 % max | 2 % |
| DiT audio output, mean of \|noise_pred_audio\|, 30 steps × 8 ranks | 0.51 % max | 2.5 % |
| LoRA sum of squares after `optimizer.step`, 30 steps × 8 ranks | 0.18 % max | 1.5 % |
| LoRA gradient norm over blocks 40 to 49, step 1, 8 ranks | 5.1 % max | 10 % |
| LoRA gradient norm over all 416 tensors, 30 steps × 8 ranks | 4.8x the baseline value, step 14, one rank (0.043 vs 0.247) | 20x, a NaN and blow-up check |
| gradient tensor count, sampled timestep, input shapes | identical | exact |

The recording code is on the `perf/minimax-h3-lora-training-verify` branch of my fork (`0391d1c`, this branch plus three commits), not in this PR; the baseline record comes from `ce9f454` plus the same three commits. The baseline itself passes the same 5,040 checks against a record of the unmodified code taken without `DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch` (closest call 0.46 of budget), so the kernel that variable selects stays inside the same tolerances.

<details>
<summary>Details: hardware, model, full command, what I tried and rejected</summary>

**Hardware.** One node, 8 × NVIDIA B200 (sm_100, 183 GB each), driver 580.126.20, CUDA 13.0; weights (62 GB) and the stage-1 cache staged to node-local NVMe. Python 3.11.16, torch 2.14.0+cu130 (cuDNN 9.24.0, triton 3.8.0), flash-attn 2.8.4 built from source for sm_100, peft 0.20.0, accelerate 1.14.0, transformers 5.16.1, DiffSynth-Studio editable from the checkout under test. Nothing was installed or removed between the two versions.

**Measurement.** Both versions on the same node with nothing else running and every GPU at 0 MiB before each run: 30 steps per process (`--dataset_repeat 240`), seed 42 + rank, two interleaved runs per version in the order baseline, this PR, baseline, this PR. Step time is the median of steps 6 to 30 on rank 0 from the script's own tqdm bar, which agrees with CUDA events around each optimizer step within 1 ms; steps 1 to 5 are warm-up. Peak memory is nvidia-smi `memory.used` sampled every 5 s, maximum over the run, identical on all 8 GPUs. Run-to-run spread is 0.03 % on the baseline and 0.04 % on this branch.

Both versions run with `DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch`. That variable is the repo's own (`docs/en/Pipeline_Usage/Environment_Variables.md`) and is not part of this PR: on this node it takes the unmodified code from 6.30 s to 3.74 s per step, because flash-attn imports in this environment and the import-time dispatch otherwise picks FlashAttention-2, while torch's SDPA dispatch selects a cuDNN sm100 kernel (17.3 ms against 61.1 ms for forward plus backward at this shape). An earlier version of this PR patched that inside the H3 model; that commit is gone, and #1687 has the measurements. The speedup this PR claims is measured on top of the variable.

**Model and job.** MiniMax-H3 FL2VA DiT: 50 blocks, hidden 5376, 56 heads × 128, FFN 14336, about 33 B parameters, bf16, frozen. LoRA rank 32 on `attn.qkv_proj, attn.out_proj, mlp.fc1, mlp.fc2`: 416 tensors, 155 M parameters, bf16. AdamW, lr 1e-4, gradient checkpointing per DiT block, 8-process DDP through accelerate (`mixed_precision: 'no'`). One sample per step per process: 124 frames at 480×832 with audio, packed into 14,912 tokens (14,430 video, 414 audio, 68 text); flow-matching MSE on video and audio. train.py does not seed, so the launcher seeds python, numpy and torch with 42 + rank. Data comes from stage 1 of the same script (`--task sft:data_process`), run once.

**The measured job**, both versions; only the checkout differs:
```
export DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch      # the setup, not the PR; exported for both versions
export DIFFSYNTH_MODEL_BASE_PATH=<weights root> DIFFSYNTH_SKIP_DOWNLOAD=True HF_HUB_OFFLINE=1
export TRITON_CACHE_DIR=<persistent> TORCHINDUCTOR_CACHE_DIR=<persistent> TORCH_EXTENSIONS_DIR=<persistent>
export PYTHONPATH=<checkout>                          # ce9f454 or this branch
export PYTHONHASHSEED=0 BENCH_SEED=42 BENCH_TRAIN_PY=<checkout>/examples/minimax_h3/model_training/train.py
accelerate launch --config_file accelerate_config_ddp8.yaml --num_processes 8 bench_train.py \
  --dataset_base_path <stage-1 cache> --data_file_keys "video,input_audio" --extra_inputs "input_audio" \
  --height 480 --width 832 --num_frames 124 --learning_rate 1e-4 --remove_prefix_in_ckpt "pipe.dit." \
  --lora_base_model dit --lora_target_modules "attn.qkv_proj,attn.out_proj,mlp.fc1,mlp.fc2" --lora_rank 32 \
  --use_gradient_checkpointing --dataset_repeat 240 --num_epochs 1 \
  --model_id_with_origin_paths "MiniMax/MiniMax-H3:FL2VA/transformer/model*.safetensors" \
  --output_path <out> --find_unused_parameters --enable_csv_log --task sft:train
```
Every flag is the recipe's; only `--dataset_repeat 240` differs, to give 30 steps per process. `bench_train.py` is a 20-line launcher that seeds with `BENCH_SEED + rank` and runs `train.py` unchanged as `__main__`. The repo ships no accelerate config for the LoRA path, so `accelerate_config_ddp8.yaml` mirrors its ZeRO-3 yaml with plain DDP and `mixed_precision: 'no'` (`bf16` there would mean autocast under MULTI_GPU, which fails FlashAttention's dtype check).

**Tried on the unmodified code and rejected** (one change at a time, on top of the variable above): `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` (3.739 s, the same step time, 0.8 GB less reserved), `TORCH_CUDNN_SDPA_DEPRIORITIZED=1` (6.557 s, +75 %), FlashAttention-4 through `DIFFSYNTH_ATTENTION_IMPLEMENTATION=flash_attention_4` (3.934 s, +5.2 %, uninstalled again). Excluded because they change the numerics or the recipe: `mixed_precision: bf16`, dropping `--use_gradient_checkpointing`, `--use_gradient_checkpointing_offload`, gradient accumulation, LoRA rank, frame count, resolution, `--customized_optimizer`.

**Traces** (torch.profiler, steps 8 to 10, all 8 ranks; open in https://ui.perfetto.dev). Neither trace is of an arm in the tables above: the baseline one is the unmodified code *without* `DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch`, and the optimized one still contains the dropped attention commit. They show the mechanism, not the numbers.
- baseline: https://drive.google.com/open?id=1LKNGMZntNh7lB4rS9hYJfF26kG8qemyX
- optimized: https://drive.google.com/open?id=1cQvISI_4vtQr7F0pX0f-grpWPDHnJZ2z

</details>

### CODE REVIEW

**Comment by TarzanZhao** [comment]
Ready for review: this PR speeds up MiniMax-H3 LoRA training in diffsynth/models/minimax_h3_dit.py, taking the training step from 6.32 s to 2.53 s (2.5x) on 8x B200.

cc @Artiprocher (merged the last change to this file, #1659) @mi804 (authored the MiniMax-H3 DiT and its refactors, #1548, #1555, #1566)

**Comment by mi804** [comment]
@TarzanZhao Thank you for your contribution and for speeding up MiniMax-H3 LoRA training! We’ll review this PR as soon as possible.

**Comment by mi804** [comment]
@TarzanZhao Thanks a lot for the detailed report and the correctness verification — the methodology is really thorough. After reading through it there are a few things I'd like to discuss:

**1. Attention backend.** DiffSynth already routes attention through `attention_forward`, controllable via `DIFFSYNTH_ATTENTION_IMPLEMENTATION`, so an override inside H3 would silently ignore the user's choice for that model only. The FA2 slowness you measured comes from `flash-attn` being installed — `DIFFSYNTH_ATTENTION_IMPLEMENTATION=torch` should already give your 3.80 s with no patch.

**2. Compile.** With `dynamic=False` and `seq_bounds` passed as Python ints, every distinct caption length becomes a new graph (H3's packed length is per-sample, with no prompt padding), so a varied dataset hits `recompile_limit` and quietly falls back to eager. The framework already has [`compile_pipeline()`](https://github.com/modelscope/DiffSynth-Studio/blob/main/docs/en/Pipeline_Usage/Accelerated_Inference.md#computation-graph-compilation) for this, `dynamic=True` by default, and H3 is already wired for it.

**3. Model-local special cases.** The patch adds quite a few mechanisms that take effect for this model alone and rely on special-case branching, and some of them may break general training compatibility. This is disastrous for the framework.

**Comment by TarzanZhao** [comment]
@mi804 Thanks for reading it so carefully. I agree with all three points and have moved this PR back to draft.

I opened it from a different angle: I wanted to show how I silently lost performance on this codebase even though I followed the installation README. If the fix should not be at the code level, the installation docs might be the right place to say it. 

I've written up my observations on the performance issues in Issue #1687.
